File: frontend/src/types/match.ts
export type ScopeMode = "companies" | "favorites"

export interface MatchScope {
  mode: ScopeMode
  company_ids: string[]
}

/**
 * One entry in an assistant turn's reasoning trace.
 *
 * `narration` steps are the model thinking out loud; `tool` steps carry the
 * call, its arguments, and the observation the model read back. They are kept
 * in one ordered array because the agent interleaves them.
 */
export interface TraceStep {
  type: "narration" | "tool"
  index: number
  content?: string
  call_id?: string | null
  name?: string | null
  label?: string
  args?: Record<string, unknown> | string | null
  ok?: boolean
  summary?: string
  observation?: string
  count?: number | null
  duration_ms?: number | null
  /** Set while the tool is still running; never persisted. */
  pending?: boolean
}

/**
 * How far a turn got.
 *
 * `running` means a live run is still writing the row and the client should
 * subscribe; `interrupted` means a backend restart stranded it, so there is
 * nothing left to subscribe to.
 */
export type TurnStatus =
  | "running"
  | "completed"
  | "stopped"
  | "failed"
  | "interrupted"

/** The complete state of a turn as of `seq` — a value, not a delta. */
export interface MatchSnapshot {
  message_id: string | null
  seq: number
  status: TurnStatus
  steps: TraceStep[]
  final_answer: string
  job_ids: string[]
}

export interface MatchMessage {
  id: string
  session_id: string
  role: "user" | "assistant"
  /** Assistant narration between tool calls — not the deliverable. */
  content: string
  /** The `final_answer` payload, rendered as the message body. */
  final_answer: string | null
  scope: MatchScope | null
  resume_id: string | null
  steps: TraceStep[]
  /**
   * Jobs the turn recommended, parsed server-side from the answer's `:job[…]`
   * markers. Kept as a record — the cards themselves render inline, so nothing
   * in the UI reads this.
   */
  job_ids: string[]
  status: TurnStatus
  /** Frames folded into this row so far — the resume anchor. */
  seq: number
  created_at: string
}

export interface Conversation {
  id: string
  title: string
  message_count: number
  created_at: string
  updated_at: string
}

export interface SendMessagePayload {
  content: string
  scope: MatchScope
  resume_id?: string | null
}

export interface ResumeSummary {
  id: string
  label: string
  filename: string
  is_default: boolean
  uploaded_at: string
}

File: backend/app/schemas/match.py
"""Request/response models for conversational job matching."""

from typing import Literal

from pydantic import BaseModel, Field


class MatchScope(BaseModel):
    """Which jobs a turn is allowed to consider.

    Exactly one mode is active. ``companies`` restricts to the named companies;
    ``favorites`` restricts to the user's saved jobs.
    """

    mode: Literal["companies", "favorites"] = "companies"
    company_ids: list[str] = Field(default_factory=list)

    def describe(self) -> str:
        """Render the scope for the agent's system prompt."""
        if self.mode == "favorites":
            return "用户的收藏岗位"
        if self.company_ids:
            return "、".join(self.company_ids) + " 的岗位"
        return "未指定范围"


class TraceStep(BaseModel):
    """One entry in an assistant turn's reasoning trace.

    ``narration`` steps hold the model's intermediate prose; ``tool`` steps hold
    the call, its arguments, and the observation the model read back.
    """

    type: Literal["narration", "tool"] = "tool"
    index: int = 0
    content: str = ""
    call_id: str | None = None
    name: str | None = None
    label: str = ""
    args: dict | str | None = None
    ok: bool = True
    summary: str = ""
    observation: str = ""
    count: int | None = None
    duration_ms: float | None = None


class MatchMessageOut(BaseModel):
    """A stored message as the UI renders it."""

    id: str
    session_id: str
    role: str
    content: str = ""
    final_answer: str | None = None
    scope: MatchScope | None = None
    resume_id: str | None = None
    steps: list[TraceStep] = Field(default_factory=list)
    job_ids: list[str] = Field(default_factory=list)
    created_at: str


class ConversationOut(BaseModel):
    """A conversation as listed in the sidebar."""

    id: str
    title: str
    message_count: int = 0
    created_at: str
    updated_at: str


class ConversationCreate(BaseModel):
    """Payload for creating a conversation."""

    title: str = ""


class ConversationUpdate(BaseModel):
    """Payload for renaming a conversation."""

    title: str = Field(min_length=1, max_length=200)


class SendMessageRequest(BaseModel):
    """A user turn."""

    content: str = Field(min_length=1, max_length=8000)
    scope: MatchScope = Field(default_factory=MatchScope)
    resume_id: str | None = None
